### PR TITLE
add variable of sigma_ieta_ieta at electron seed level

### PR DESCRIPTION
**Pull request by doanhien**

### CODE REVIEW

**Comment by lgray** [comment]
@doanhien Can you please link here some summary plots and fill a description of the pull request?

Thanks.

**Comment by doanhien** [comment]
Hi Lindsey,

What summary plots do you mention?

Best,
Hien.

**Comment by lgray** [comment]
@doanhien The ones you have shown previously by mail.

File: RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc [hunk 1]
```diff
@@ -83,6 +88,13 @@ ElectronSeedProducer::ElectronSeedProducer( const edm::ParameterSet& iConfig )
      maxHEndcaps_=conf_.getParameter<double>("maxHEndcaps") ;
    }
 
+  applySigmaIEtaIEtaCut_ = conf_.getParameter<bool>("applySigmaIEtaIEtaCut");
+  if (applySigmaIEtaIEtaCut_)
+    {
+      maxSigmaIEtaIEtaBarrel_ = conf_.getParameter<double>("maxSigmaIEtaIEtaBarrel");
+      maxSigmaIEtaIEtaEndcaps_ = conf_.getParameter<double>("maxSigmaIEtaIEtaEndcaps");
+    }
```
Comment (lgray) [inline RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc]: This is only meant to be applied for heavy ion data, correct?

File: RecoEgamma/EgammaElectronProducers/python/ecalDrivenElectronSeedsParameters_cff.py [hunk 2]
```diff
@@ -34,7 +34,11 @@
     # H/E towers
     hcalTowers = cms.InputTag("towerMaker"),
     hOverEPtMin = cms.double(0.),
-    
+
+    applySigmaIEtaIEtaCut = cms.bool(True),
+    # sigma_ietaieta
+    maxSigmaIEtaIEtaBarrel = cms.double(0.5),
+    maxSigmaIEtaIEtaEndcaps = cms.double(0.5),
```
Comment (lgray) [inline RecoEgamma/EgammaElectronProducers/python/ecalDrivenElectronSeedsParameters_cff.py]: If so, it should be turned off here and this module should be customized such that it is turned on for HI reconstruction.

**Comment by doanhien** [comment]
Hi Lindsey,

This is the link of plots applied sigma_ieta_ieta cut.

Best,
Hien.

**Comment by doanhien** [comment]
Oops, sorry, I forgot to add link in previous mail.

https://twiki.cern.ch/twiki/pub/CMS/PhotonAnalyses2015/20150616.pdf

Best,
Hien.

File: RecoEgamma/EgammaElectronProducers/python/ecalDrivenElectronSeedsParameters_cff.py (hunk 2 above)
Comment (doanhien) [inline RecoEgamma/EgammaElectronProducers/python/ecalDrivenElectronSeedsParameters_cff.py]: Hi Lindsey,



On Tue, Jul 14, 2015 at 3:08 PM, Lindsey Gray <notifications@github.com>
wrote:

> In
> RecoEgamma/EgammaElectronProducers/python/ecalDrivenElectronSeedsParameters_cff.py
> <https://github.com/cms-sw/cmssw/pull/10194#discussion_r34565363>:
>
> > @@ -34,7 +34,11 @@
> >      # H/E towers
> >      hcalTowers = cms.InputTag("towerMaker"),
> >      hOverEPtMin = cms.double(0.),
> > -
> > +
> > +    applySigmaIEtaIEtaCut = cms.bool(True),
> > +    # sigma_ietaieta
> > +    maxSigmaIEtaIEtaBarrel = cms.double(0.5),
> > +    maxSigmaIEtaIEtaEndcaps = cms.double(0.5),
>
> If so, it should be turned off here and this module should be customized
> such that it is turned on for HI reconstruction.
>

Do you mean I have to customize this module for HI reconstruction?

Thanks,
Hien.

Comment (lgray) [inline RecoEgamma/EgammaElectronProducers/python/ecalDrivenElectronSeedsParameters_cff.py]: Yes, there should be an appropriate file to add the necessary cuts to the HI workflow.
@richard-cms should know where you need to make this customization.

**Comment by richard-cms** [comment]
@lgray , @doanhien 

This PR is meant to allow a new cut at the electron seeding step for use by HI. The boolean "applySigmaIEtaIEtaCut" should default to False (and the cfi file should be updated to reflect this). 

For adding this to the HI workflow, I'd like to see this PR merged first and then I will make a separate PR modifying the HI workflow, so that we can test different cut values on a variety of samples.

**Comment by doanhien** [comment]
@lgray, @richard-cms 

Hi,

I already change the boolean "applySigmaIEtaIEtaCut" to False as default value.

**Comment by lgray** [comment]
OK That's fine for me.

Then no changes expected/observed for this PR.
Can you please mark that in the main conversation?

Thanks,
-L

**Comment by doanhien** [comment]
@cmsbuild 
Hi,

I already changed the boolean "applySigmaIEtaIEtaCut" to False as default value.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by doanhien** [comment]
Hi,

Thanks for your testing. This is my mistake. These collections: 

'ebRecHitCollection'
 'eeRecHitCollection'

are used in HI, not in pp reconstruction. 
Is it possible if I create a configuration file used only for HI (e.g: ecalDrivenElectronSeeds_HI_cfi.py?

**Comment by richard-cms** [comment]
@doanhien The errors are because you have added new parameters (ebRecHitCollection and eeRecHitCollection) without also adding them to the fillDescriptions method. 

Additionally I think that you should move the code which grabs the rechit collections into the block protected by the new boolean, so that the collections are grabbed only if the boolean is true. I'll add line comments to show you what I mean.

File: RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc [hunk 3]
```diff
@@ -323,6 +356,7 @@ ElectronSeedProducer::fillDescriptions(edm::ConfigurationDescriptions& descripti
     psd0.add<edm::InputTag>("measurementTrackerEvent",edm::InputTag("MeasurementTrackerEvent"));
     psd0.add<edm::InputTag>("vertices",edm::InputTag("offlinePrimaryVerticesWithBS"));
     psd0.add<bool>("applyHOverECut",true);
+    psd0.add<bool>("applySigmaIEtaIEtaCut", false);
```
Comment (richard-cms) [inline RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc]: You need to add eeRecHitCollection and ebRecHitCollection to the appropriate section in here.

File: RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc [hunk 4]
```diff
@@ -94,6 +107,9 @@ ElectronSeedProducer::ElectronSeedProducer( const edm::ParameterSet& iConfig )
   matcher_ = new ElectronSeedGenerator(conf_,esg_tokens) ;
 
   //  get collections from config'
+  ebRecHitCollection_ = consumes<EcalRecHitCollection> (iConfig.getParameter<edm::InputTag>("ebRecHitCollection"));
+  eeRecHitCollection_ = consumes<EcalRecHitCollection> (iConfig.getParameter<edm::InputTag>("eeRecHitCollection"));
```
Comment (richard-cms) [inline RecoEgamma/EgammaElectronProducers/plugins/ElectronSeedProducer.cc]: I think you should add an "if (applySigmaIEtaIEtaCut_ == true)" around these consumes statements.

**Comment by doanhien** [comment]
@richard-cms 
Thanks for your comment. I just add them to .cc file.

**Comment by doanhien** [comment]
@cmsbuild 

Hi,

I already added something in .cc file as Alex's suggestion to fix the problem of 'ebRechitCollection' and 'eeRechitCollection'.
Can you please test it?

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by smuzaffar** [comment]
test

**Comment by richard-cms** [comment]
I spoke with @doanhien on Skype about how to test before pushing and also how to fix this particular error.

**Comment by lgray** [comment]
Thanks Alex!

**Comment by doanhien** [comment]
This particular was fixed with help from Alex. Test was done locally without errors.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #10194 d7426aca48d097883cd6f6b950bd874d4dd89708

Added a sigma_ieta_ieta variable at electron seed level for use with HI workflows. The cut is disabled by default, so this PR should not change monitored quantities. A later PR may enable this cut after testing has established the desired cut value.

The code changes are satisfactory, and Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10194/6251/summary.html) against baseline CMSSW_7_6_X_2015-07-14-1100 show no significant differences, as expected.

**Comment by davidlange6** [comment]
+1

**Comment by richard-cms** [comment]
@doanhien Please make sure to backport this to 7_5_X as well.

**Pull request merged**